File: sdks/local-rp/python/tests/test_conformance_claims.py
"""claims.json: Claim wire encoding and claim-signature verification.

THE TRAP THIS FILE EXISTS TO CATCH: `Claim.claim_value` is CBOR bytes (bstr,
major type 2), never a text string (tstr, major type 3) -- both on the wire
and inside the signed payload. An SDK that wires it as text passes its own
self-tests perfectly (sign-wrong/verify-wrong is self-consistent) and only
cross-implementation vectors expose the bug. Every case here runs through the
SDK's own claim verification path (`linkkeys_local_rp/claims.py`), the same
module `complete_local_login` uses.
"""

from datetime import datetime, timezone

import pytest
from conftest import hex_decode, load_vector

from linkkeys_local_rp import claims as claims_mod
from linkkeys_local_rp.generated import codec as _codec  # noqa: F401  (attaches to_cbor/from_cbor)
from linkkeys_local_rp.generated.types import Claim, ClaimSignature, DomainPublicKey

# All fixture timestamps are fixed 2026-01-01 constants with far-future
# (2126) or absent expiry; any fixed "now" between issuance and 2126 works
# for every case here.
NOW = datetime(2026, 1, 2, tzinfo=timezone.utc)

_ERROR_KIND_BY_CLASS = {
    claims_mod.SignatureInvalid: "signature_invalid",
    claims_mod.KeyNotFound: "key_not_found",
}


@pytest.fixture(scope="module")
def claims_vectors():
    return load_vector("claims.json")


def _claim_signature_from_case(s: dict) -> ClaimSignature:
    return ClaimSignature(
        domain=s["domain"],
        signed_by_key_id=s["signed_by_key_id"],
        signature=hex_decode(s["signature_hex"]),
    )


def _claim_from_case(case: dict) -> Claim:
    c = case["claim"]
    return Claim(
        claim_id=c["claim_id"],
        user_id=c["user_id"],
        claim_type=c["claim_type"],
        claim_value=hex_decode(c["claim_value_hex"]),
        signatures=[_claim_signature_from_case(s) for s in c["signatures"]],
        attested_at=c["attested_at"],
        created_at=c["created_at"],
        expires_at=c["expires_at"],
        revoked_at=c["revoked_at"],
    )


def _domain_public_key(k: dict) -> DomainPublicKey:
    return DomainPublicKey(
        key_id=k["key_id"],
        public_key=hex_decode(k["public_key_hex"]),
        fingerprint=k["fingerprint_hex"],
        algorithm=k["algorithm"],
        key_usage=k["key_usage"],
        created_at=k["created_at"],
        expires_at=k["expires_at"],
        revoked_at=k["revoked_at"],
        signed_by_key_id=None,
        key_signature=None,
    )


def _domain_key_sets(entries: list) -> list:
    """Group flat `domain_keys` entries by their `domain` field into the
    `DomainKeySet` list `verify_claim`/`verify_claim_signatures` expect."""
    by_domain: dict = {}
    for k in entries:
        by_domain.setdefault(k["domain"], []).append(_domain_public_key(k))
    return [claims_mod.DomainKeySet(domain=d, keys=ks) for d, ks in by_domain.items()]


def _error_kind(exc: Exception) -> str:
    for cls, name in _ERROR_KIND_BY_CLASS.items():
        if isinstance(exc, cls):
            return name
    raise AssertionError(f"unmapped claim error type: {type(exc)!r}: {exc}")


def test_registry_constants(claims_vectors):
    assert claims_vectors["tag"] == claims_mod.CLAIM_PAYLOAD_TAG == "linkkeys-claim-v2"


def test_positive_cases_round_trip_and_verify(claims_vectors):
    """Byte-exact wire round trip, exact per-signature signed-payload bytes,
    and full `verify_claim` success -- for every positive case, through the
    SDK's own decode/verify path."""
    cases = claims_vectors["cases"]
    assert len(cases) == 3
    domain_keys = _domain_key_sets(claims_vectors["domain_keys"])

    for case in cases:
        wire = hex_decode(case["claim_cbor_hex"])
        expected = _claim_from_case(case)

        decoded = claims_mod.decode_claim(wire)
        assert decoded == expected, case["name"]
        assert isinstance(decoded.claim_value, bytes), case["name"]

        # Byte-exact re-encode via the generated codec.
        assert decoded.to_cbor() == wire, case["name"]

        # Exact per-signature signed-payload bytes (the 8-element tag-first
        # array, subject as the single '@'-joined string).
        for sig_case, sig in zip(case["claim"]["signatures"], decoded.signatures):
            payload = claims_mod.claim_sign_payload(
                decoded.claim_id,
                decoded.claim_type,
                decoded.claim_value,
                decoded.user_id,
                case["subject_domain"],
                sig.domain,
                decoded.expires_at,
                decoded.attested_at,
            )
            assert payload == hex_decode(sig_case["signed_payload_cbor_hex"]), case["name"]

        assert case["expected_valid"] is True
        claims_mod.verify_claim(decoded, case["subject_domain"], domain_keys, NOW)


def test_claim_non_utf8_binary_value_is_not_valid_utf8(claims_vectors):
    """The discriminator case: its value bytes are not valid UTF-8, so a
    tstr-based codec could not even represent them."""
    case = next(c for c in claims_vectors["cases"] if c["name"] == "claim_non_utf8_binary_value")
    value = hex_decode(case["claim"]["claim_value_hex"])
    with pytest.raises(UnicodeDecodeError):
        value.decode("utf-8")


def test_claim_value_as_cbor_text_rejected(claims_vectors):
    """The wire trap: byte-identical to claim_utf8_text_value's CBOR except
    claim_value is encoded as a CBOR text string (major type 3) instead of a
    byte string (major type 2). A strict bstr codec must refuse to decode
    this at all -- not merely mis-verify it."""
    cases = claims_vectors["decode_negative_cases"]
    assert len(cases) == 1
    case = cases[0]
    assert case["expected_decode_ok"] is False
    assert case["name"] == "claim_value_as_cbor_text_rejected"

    wire = hex_decode(case["claim_cbor_hex"])
    with pytest.raises(_codec.CsilDecodeError):
        claims_mod.decode_claim(wire)


def test_verification_negative_cases(claims_vectors):
    cases = claims_vectors["negative_cases"]
    assert len(cases) == 4
    default_domain_keys = _domain_key_sets(claims_vectors["domain_keys"])

    for case in cases:
        wire = hex_decode(case["claim_cbor_hex"])
        claim = claims_mod.decode_claim(wire)
        domain_keys = _domain_key_sets(case["domain_keys"]) if "domain_keys" in case else default_domain_keys

        with pytest.raises(claims_mod.ClaimError) as exc_info:
            claims_mod.verify_claim(claim, case["subject_domain"], domain_keys, NOW)
        assert _error_kind(exc_info.value) == case["expected_error"], case["name"]


def test_ticket_redemption_response_round_trips_and_verifies(claims_vectors):
    """The wire message `complete_local_login` actually consumes Claims
    from. Decoding must reproduce each claim byte-exactly (claim_value as
    raw bytes), re-encoding must reproduce response_cbor_hex, and every
    embedded claim's signatures must verify."""
    trr = claims_vectors["ticket_redemption_response"]
    wire = hex_decode(trr["response_cbor_hex"])

    decoded = claims_mod.decode_ticket_redemption_response(wire)

    assert decoded.user_id == trr["user_id"]
    assert decoded.user_domain == trr["user_domain"]
    assert decoded.ticket_expires_at == trr["ticket_expires_at"]

    expected_claims = [_claim_from_case(c) for c in claims_vectors["cases"]]
    assert decoded.claims == expected_claims
    for claim in decoded.claims:
        assert isinstance(claim.claim_value, bytes)

    # Byte-exact re-encode.
    assert decoded.to_cbor() == wire

    domain_keys = _domain_key_sets(claims_vectors["domain_keys"])
    for claim in decoded.claims:
        claims_mod.verify_claim(claim, decoded.user_domain, domain_keys, NOW)


def test_ticket_redemption_response_with_text_claim_value_is_rejected(claims_vectors):
    """The bstr/tstr check must apply to every claim embedded in a
    `LocalRpTicketRedemptionResponse` -- the path `redeem_claim_ticket`
    actually uses -- so a compromised/malicious IDP smuggling a text-encoded
    claim_value inside a real redemption response is rejected. The generated
    codec is now strict at every decode layer, so the malicious wire bytes
    are spliced together at the raw-CBOR level (as an attacker would), not
    via the typed API (which can no longer represent them)."""
    trr = claims_vectors["ticket_redemption_response"]
    bad_claim_wire = hex_decode(claims_vectors["decode_negative_cases"][0]["claim_cbor_hex"])

    # Sanity: the bad claim can't even be decoded standalone anymore -- the
    # generated codec itself rejects the text-typed claim_value.
    with pytest.raises(_codec.CsilDecodeError):
        Claim.from_cbor(bad_claim_wire)

    # Splice the raw bad-claim tree into an otherwise-valid response tree and
    # re-encode with the generic (untyped) CBOR encoder.
    response_tree = _codec.cbor_decode(hex_decode(trr["response_cbor_hex"]))
    bad_claim_tree = _codec.cbor_decode(bad_claim_wire)
    response_tree["claims"].append(bad_claim_tree)
    wire = _codec.cbor_encode(response_tree)

    with pytest.raises(_codec.CsilDecodeError):
        claims_mod.decode_ticket_redemption_response(wire)

File: sdks/local-rp/ruby/test/test_conformance_claims.rb
# frozen_string_literal: true

require_relative 'test_helper'

class TestConformanceClaims < Minitest::Test
  include ConformanceHelper

  # `now` for signature-quorum/expiry checks -- coincides with every
  # fixture's attested_at/created_at, and is safely before claim 1's
  # far-future expires_at (2126) and after nothing else in the file has an
  # expiry.
  NOW = LinkkeysLocalRp::Timeutil.parse_rfc3339('2026-01-01T00:00:00+00:00')

  ERROR_CLASS_FOR = {
    'signature_invalid' => LinkkeysLocalRp::Claims::SignatureInvalid,
    'key_not_found' => LinkkeysLocalRp::Claims::KeyNotFound
  }.freeze

  def build_domain_key(k)
    LinkkeysLocalRp::Types::DomainPublicKey.new(
      key_id: k['key_id'],
      public_key: ConformanceHelper.hex(k['public_key_hex']),
      fingerprint: k['fingerprint_hex'],
      algorithm: k['algorithm'],
      key_usage: k['key_usage'],
      created_at: k['created_at'],
      expires_at: k['expires_at'],
      revoked_at: k['revoked_at'],
      signed_by_key_id: nil,
      key_signature: nil
    )
  end

  # claims.json's domain_keys is a flat list (every entry happens to share
  # one domain in this fixture); group it into the DomainKeySet list shape
  # Claims.verify_claim / complete_local_login actually consume.
  def build_domain_key_sets(list)
    list.group_by { |k| k['domain'] }.map do |domain, keys|
      LinkkeysLocalRp::Claims::DomainKeySet.new(domain: domain, keys: keys.map { |k| build_domain_key(k) })
    end
  end

  def default_domain_key_sets
    @default_domain_key_sets ||= build_domain_key_sets(claims_vector['domain_keys'])
  end

  def build_claim_signature(s)
    LinkkeysLocalRp::Types::ClaimSignature.new(
      domain: s['domain'], signed_by_key_id: s['signed_by_key_id'], signature: ConformanceHelper.hex(s['signature_hex'])
    )
  end

  def build_claim(claim_h)
    LinkkeysLocalRp::Types::Claim.new(
      claim_id: claim_h['claim_id'],
      user_id: claim_h['user_id'],
      claim_type: claim_h['claim_type'],
      claim_value: ConformanceHelper.hex(claim_h['claim_value_hex']),
      signatures: claim_h['signatures'].map { |s| build_claim_signature(s) },
      attested_at: claim_h['attested_at'],
      created_at: claim_h['created_at'],
      expires_at: claim_h['expires_at'],
      revoked_at: claim_h['revoked_at']
    )
  end

  def find_domain_public_key(domain_keys_hashes, key_id)
    entry = domain_keys_hashes.find { |k| k['key_id'] == key_id }
    raise "fixture key #{key_id} not found" if entry.nil?

    ConformanceHelper.hex(entry['public_key_hex'])
  end

  # --- positive cases -----------------------------------------------------

  def test_registry_constants
    assert_equal claims_vector['tag'], LinkkeysLocalRp::Claims::CLAIM_PAYLOAD_TAG
  end

  def test_positive_case_wire_round_trip
    claims_vector['cases'].each do |c|
      claim = build_claim(c['claim'])
      encoded = claim.to_cbor.unpack1('H*')
      assert_equal c['claim_cbor_hex'], encoded, "encode mismatch: #{c['name']}"

      decoded = LinkkeysLocalRp::Types::Claim.from_cbor(ConformanceHelper.hex(c['claim_cbor_hex']))
      reencoded = decoded.to_cbor.unpack1('H*')
      assert_equal c['claim_cbor_hex'], reencoded, "decode->encode round-trip mismatch: #{c['name']}"

      # The bstr/tstr trap: claim_value must survive the round trip as raw
      # bytes (ASCII-8BIT), never coerced into a UTF-8 text string, even for
      # the case whose bytes happen to be valid UTF-8.
      assert_equal ::Encoding::ASCII_8BIT, decoded.claim_value.encoding, "claim_value not decoded as bstr: #{c['name']}"
      assert_equal ConformanceHelper.hex(c['claim']['claim_value_hex']), decoded.claim_value, "claim_value bytes mismatch: #{c['name']}"
    end
  end

  # Independent of the SDK's own claim_sign_payload construction: verify
  # each fixture's signed_payload_cbor_hex/signature_hex directly against
  # the domain's public key using the raw Ed25519 primitive, so a correct
  # verify_claim run below can't be hiding a payload-construction bug that
  # happens to be self-consistent.
  def test_signed_payload_independent_ed25519_verification
    claims_vector['cases'].each do |c|
      c['claim']['signatures'].each do |s|
        payload = ConformanceHelper.hex(s['signed_payload_cbor_hex'])
        signature = ConformanceHelper.hex(s['signature_hex'])
        public_key = find_domain_public_key(claims_vector['domain_keys'], s['signed_by_key_id'])

        LinkkeysLocalRp::Crypto.verify_with_algorithm(
          LinkkeysLocalRp::Crypto::SigningAlgorithm::ED25519, payload, signature, public_key
        )
      rescue LinkkeysLocalRp::Crypto::VerificationFailed => e
        flunk "#{c['name']} / #{s['signed_by_key_id']}: independent Ed25519 verification failed: #{e.message}"
      end
    end
  end

  def test_positive_cases_verify_claim
    claims_vector['cases'].each do |c|
      claim = build_claim(c['claim'])
      LinkkeysLocalRp::Claims.verify_claim(claim, c['subject_domain'], default_domain_key_sets, NOW)
    rescue LinkkeysLocalRp::Claims::Error => e
      flunk "#{c['name']}: expected verify_claim to succeed, raised #{e.class}: #{e.message}"
    end
  end

  # --- decode negatives ----------------------------------------------------

  def test_decode_negative_cases
    claims_vector['decode_negative_cases'].each do |c|
      refute c['expected_decode_ok'], "fixture bug: #{c['name']} should have expected_decode_ok=false"

      assert_raises(LinkkeysLocalRp::Cbor::DecodeError, "#{c['name']}: decode should have been rejected") do
        LinkkeysLocalRp::Types::Claim.from_cbor(ConformanceHelper.hex(c['claim_cbor_hex']))
      end
    end
  end

  # --- verification negatives ----------------------------------------------

  def test_verification_negative_cases
    claims_vector['negative_cases'].each do |c|
      claim = LinkkeysLocalRp::Types::Claim.from_cbor(ConformanceHelper.hex(c['claim_cbor_hex']))
      domain_key_sets = c['domain_keys'] ? build_domain_key_sets(c['domain_keys']) : default_domain_key_sets
      expected_class = ERROR_CLASS_FOR.fetch(c['expected_error']) do
        flunk "#{c['name']}: unmapped expected_error #{c['expected_error']}"
      end

      assert_raises(expected_class, "#{c['name']}: expected #{c['expected_error']}") do
        LinkkeysLocalRp::Claims.verify_claim(claim, c['subject_domain'], domain_key_sets, NOW)
      end
    end
  end

  # --- ticket redemption response -------------------------------------------

  def test_ticket_redemption_response_round_trip_and_claim_verification
    trr = claims_vector['ticket_redemption_response']
    response = LinkkeysLocalRp::Types::LocalRpTicketRedemptionResponse.from_cbor(
      ConformanceHelper.hex(trr['response_cbor_hex'])
    )

    assert_equal trr['user_id'], response.user_id
    assert_equal trr['user_domain'], response.user_domain
    assert_equal trr['ticket_expires_at'], response.ticket_expires_at
    assert_equal claims_vector['cases'].length, response.claims.length

    reencoded = response.to_cbor.unpack1('H*')
    assert_equal trr['response_cbor_hex'], reencoded, 'response round-trip mismatch'

    # Decoding without verifying fails the point (README's own words) --
    # verify every embedded claim's signatures through the exact path
    # complete_local_login uses.
    response.claims.each_with_index do |claim, i|
      assert_equal ::Encoding::ASCII_8BIT, claim.claim_value.encoding, "claims[#{i}].claim_value not bstr"
      LinkkeysLocalRp::Claims.verify_claim(claim, response.user_domain, default_domain_key_sets, NOW)
    rescue LinkkeysLocalRp::Claims::Error => e
      flunk "claims[#{i}] (#{claim.claim_id}): expected verify_claim to succeed, raised #{e.class}: #{e.message}"
    end
  end
end
